code/iaas/ha/src/main/java/io/cattle/platform/ha/monitor/impl/PingInstancesMonitorImpl.java

```diff
@@ -115,36 +112,49 @@ public boolean evaluate(ProcessState state, ProcessInstance processInstance, Pro
         });
     }
 
-    protected Set<String> getInstances(Ping ping) {
+    protected ReportedInstances getInstances(Ping ping) {
         PingData data = ping.getData();
-        if (data == null || ping.getResourceId() == null) {
+
+        if ( data == null || ping.getResourceId() == null ) {
             return null;
         }
 
         List<Map<String, Object>> resources = data.getResources();
-        if (resources == null || !ping.getOption(Ping.INSTANCES)) {
+        if ( resources == null || !ping.getOption(Ping.INSTANCES) ) {
             return null;
         }
 
-        Set<String> instances = new HashSet<String>();
+        ReportedInstances reportedInstances = new ReportedInstances();
 
-        for (Map<String, Object> resource : resources) {
+        for ( Map<String, Object> resource : resources ) {
             Object state = resource.get(ObjectMetaDataManager.STATE_FIELD);
             Object type = resource.get(ObjectMetaDataManager.TYPE_FIELD);
             Object uuid = resource.get(ObjectMetaDataManager.UUID_FIELD);
+            Object externalId = resource.get("dockerId");
 
-            if (!InstanceConstants.TYPE.equals(type) || !InstanceConstants.STATE_RUNNING.equals(state) || uuid == null) {
+            if ( !InstanceConstants.TYPE.equals(type) || !InstanceConstants.STATE_RUNNING.equals(state) ) {
                 continue;
             }
 
-            instances.add(uuid.toString());
+            if ( uuid != null ) {
+                reportedInstances.byUuid.add(uuid.toString());
+            }
+
+            if ( externalId != null ) {
+                reportedInstances.byExternalId.add(externalId.toString());
+            }
         }
 
-        return instances;
+        return reportedInstances;
+    }
+
+    protected class ReportedInstances {
+        Set<String> byUuid = new HashSet<String>();
+        Set<String> byExternalId = new HashSet<String>();
     }
 
-    protected Set<String> load(Long agentId) {
-        return monitorDao.getHosts(agentId);
+    protected Map<String, String> load(Long agentId) {
```

Comment:
Autoboxing! :) should either check that the agentId is not null, or change the method signature to accept primitive long as thats what MonitorDao expects.